### PR TITLE
Components core

### LINKED ISSUE #3: Adicionar classe Object/GameObject
Classe base para os objetos do jogo.

Precisa conter o método  ```Update()``` e o método ```Render()``` Para chamada na lista do GameSystem.

### PR DESCRIPTION
**Pull request by GoliasVictor**
## Systems and Scenes
The concept of Systems (`ISystem`) was added, where each system is updated at all times, passing the context of each system as a parameter for the update.

Scenes (`IScene`) were also added, which are systems that need to render something, where each frame is rendered and updated.

The order in which the systems will be updated will be according to the order in which they are in the `GameSystem` array.
## Entities
The `IEntity` interface was created, which represents an entity within the grid. It has to carry with it a position property, which is a `GridVec2` that represents its position on the map. Other methods of the interface are basically for the systems to use.

## Map

The map was created, where it has its layers, a fixed object layer, and a movable object layer. The fixed object layer is a two-dimensional array, where it will be populated by objects that will not be changed in position (Wall and floor). And the layer of movable objects, which is a list of entities. There is the `AllEntities` property that returns all entities, both fixed and movable, and the `Player` property that returns the player on the map.

## Test
The system was tested by placing a player on the screen, moving him and using a box that he can move.
Then basic forms of collision between entities were implemented. And a player movement system was added.

Note: the way movement and collision are currently implemented is a rough way, which needs to be improved. I didn't do this in more depth because it's not the goal of the PR.

### CODE REVIEW

File: src/Jogo/Program.cs [hunk 1]
```diff
@@ -1,19 +1,24 @@
...
+    public const int DefaultWindowWidth = 800;
+    public const int DefaultWindowHeight = 600;
     private const string DefaultWindowName = "Elements";
+    public const int TileSize = 30;
 
     private static Color ClearColor = Color.DarkGray;
     private static int TargetFPS = 60;
-
-
+    private static IScene ActualScene;
```
Comment (grandeisa) [inline src/Jogo/Program.cs]: Mudar nome da variável para ```CurrentScene```, 'actual' não equivale ao termo 'atual' em português.

File: src/Jogo/Models/IEntity.cs [hunk 2]
```diff
@@ -0,0 +1,5 @@
+using System.Numerics;
+
+interface IEntity{
+	void Render(LevelScene level, int x, int y);
```
Comment (grandeisa) [inline src/Jogo/Models/IEntity.cs:19]: Não faz sentido o método Render carregar a posição do objeto sendo que é responsabilidade dele ter essa informação.

**Review by grandeisa — CHANGES_REQUESTED** [review]

File: src/Jogo/Models/TileVec2.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+struct TileVec2 {
```
Comment (vitordsp) [inline src/Jogo/Models/TileVec2.cs]: tem algum motivo pra isso aqui chama TileVec[2] ?

File: src/Jogo/Scripts/PlayerMovingSystem.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
+using Raylib_cs;
+
+class PlayerMovingSystem : ISystem<LevelScene>
```
Comment (vitordsp) [inline src/Jogo/Scripts/PlayerMovingSystem.cs]: N seria melhor PlayerMovementSystem?

File: src/Jogo/Models/TileVec2.cs (hunk 3 above)
Comment (GoliasVictor) [reply src/Jogo/Models/TileVec2.cs]: Foi mudado para grid, mas é porque é um vetor de tamanho 2 na grid

File: src/Jogo/Scripts/PlayerMovingSystem.cs (hunk 4 above)
Comment (GoliasVictor) [reply src/Jogo/Scripts/PlayerMovingSystem.cs]: Meu ingles não é muito bom, vo mudar

**Review by grandeisa — APPROVED** [review]

**Review by vitordsp — APPROVED** [review]
ta top

**Pull request merged**